File: Digit-Core/core-services/digit-service/src/main/java/org/egov/digit/util/PasswordGenerator.java
package org.egov.digit.util;

import java.security.SecureRandom;

public class PasswordGenerator {

    private static final String LOWERCASE = "abcdefghijklmnopqrstuvwxyz";
    private static final String UPPERCASE = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
    private static final String NUMBERS = "0123456789";
    private static final String SYMBOLS = "!@#$%^&*()_+-=[]{}|;:,.<>?";

    private static final SecureRandom random = new SecureRandom();

    /**
     * Generate a simple alphanumeric password (default: 10 characters)
     */
    public static String generateSimplePassword() {
        return generateSimplePassword(10);
    }

    /**
     * Generate a simple alphanumeric password with specified length
     */
    public static String generateSimplePassword(int length) {
        String characters = LOWERCASE + UPPERCASE + NUMBERS;
        return generatePassword(length, characters);
    }

    /**
     * Generate a complex password with all character types
     */
    public static String generateComplexPassword(int length) {
        return generatePassword(length, true, true, true, true);
    }

    /**
     * Generate a password with custom character requirements
     */
    public static String generatePassword(int length, boolean includeUppercase, boolean includeLowercase,
                                          boolean includeNumbers, boolean includeSymbols) {
        StringBuilder charSet = new StringBuilder();

        if (includeLowercase) charSet.append(LOWERCASE);
        if (includeUppercase) charSet.append(UPPERCASE);
        if (includeNumbers) charSet.append(NUMBERS);
        if (includeSymbols) charSet.append(SYMBOLS);

        if (charSet.length() == 0) {
            charSet.append(LOWERCASE).append(NUMBERS);
        }

        return generatePassword(length, charSet.toString());
    }

    /**
     * Generate password from given character set
     */
    private static String generatePassword(int length, String characterSet) {
        if (length < 1) {
            throw new IllegalArgumentException("Password length must be at least 1");
        }

        StringBuilder password = new StringBuilder(length);
        for (int i = 0; i < length; i++) {
            int index = random.nextInt(characterSet.length());
            password.append(characterSet.charAt(index));
        }

        return password.toString();
    }
}

File: digit-be/utils/passwordGenerator.js
const crypto = require('crypto');

/**
 * Generate a random secure password
 * @param {number} length - Length of the password (default: 12)
 * @param {object} options - Options for password generation
 * @returns {string} Generated password
 */
const generatePassword = (length = 12, options = {}) => {
  const {
    includeUppercase = true,
    includeLowercase = true,
    includeNumbers = true,
    includeSymbols = true
  } = options;

  let charset = '';
  const lowercase = 'abcdefghijklmnopqrstuvwxyz';
  const uppercase = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ';
  const numbers = '0123456789';
  const symbols = '!@#$%^&*()_+-=[]{}|;:,.<>?';

  if (includeLowercase) charset += lowercase;
  if (includeUppercase) charset += uppercase;
  if (includeNumbers) charset += numbers;
  if (includeSymbols) charset += symbols;

  if (charset.length === 0) {
    throw new Error('At least one character set must be enabled');
  }

  let password = '';
  const randomBytes = crypto.randomBytes(length);

  for (let i = 0; i < length; i++) {
    const randomIndex = randomBytes[i] % charset.length;
    password += charset[randomIndex];
  }

  return password;
};

/**
 * Generate a simple alphanumeric password (easier to type)
 * @param {number} length - Length of the password (default: 10)
 * @returns {string} Generated password
 */
const generateSimplePassword = (length = 10) => {
  return generatePassword(length, {
    includeUppercase: true,
    includeLowercase: true,
    includeNumbers: true,
    includeSymbols: false
  });
};

module.exports = {
  generatePassword,
  generateSimplePassword
};
